src/WhoWroteThat.ts

```diff
@@ -0,0 +1,152 @@
+                return this.rec(declaration, commits, 0)
+                    .then(owners => ({ declaration, owners }));
+            }).then(transformResult);
+        });
+    }
+
+    line(line: number): Promise<CodeOwners> {
+        return Git.commitsAffectingFile(this.filePath).then(commits => {
+            return readFile(this.filePath).then(sourceCode => {
+                const rootNode = this.parser.parse(sourceCode);
+                const declaration =
+                    this.parser.findDeclarationByLine(rootNode, line);
+                return this.rec(declaration, commits, 0)
+                    .then(owners => ({ declaration, owners }));
+            }).then(transformResult);
+        });
+    }
+
+    private rec(
+        decl: Declaration,
+        commits: string[],
+        commitIndex: number
+    ): Promise<Owner[]> {
+        if (
+            this.depth && commitIndex >= this.depth ||
+            commitIndex >= commits.length
+        ) return new Promise(resolve => resolve([]));
+
+        return Git.readFileAtCommit(this.filePath, commits[commitIndex])
+            .then(sourceCodeAtCommit => {
+                const node = this.parser.parse(sourceCodeAtCommit);
+                const spans = this.parser.findSpans(node, decl);
+
+                // if current commit does not contain declaration
+                // assume no earlier commit contains declaration
+                if (spans.length == 0) return [];
+
+                return Promise.all(
+                    spans.map(span => Git.getOwnerOfCommit(
+                        this.filePath,
+                        commits[commitIndex],
+                        span
+                    ))
+                )
+                    .then(mergeDuplicateOwners)
+                    .then(owners => {
+                        return this.rec(decl, commits, commitIndex + 1)
+                            .then(newOwners => mergeDuplicateOwners([
+                                ...owners,
+                                ...scale(
+                                    this.strategy,
+                                    newOwners,
+                                    weight(this.strategy, owners)
+                                )
+                            ]));
+                    });
+            }).catch(() => {
+                // file could not be read at HEAD~commitIndex
+                return [];
+            });
+    }
+}
+
+const transformResult = (result: CodeOwners): CodeOwners => ({
+    ...result,
+    owners: squish(
+        result.owners
+            .sort((a, b) => a.score < b.score ? 1 : -1)
+    ).filter(owner => owner.score > 0)
+});
+
+const scale = (
+    strategy: Strategy,
+    owners: Owner[],
+    weight: number
+): Owner[] => {
+    if (strategy === Strategy.WEIGHTED_LINES)
+        return owners.map(owner => ({
+            ...owner,
+            score: owner.score * (weight - weight / 4) / weight
```

Comment:
why especially 4?

Reply:
This is just how this strategy is defined. I'm going to document it with #57